### PR TITLE
deps: validator@5.1.0

### LINKED ISSUE #6462: Update server side validator.js dependency, and fix deprecation warnings
Once we update to the latest version of validator, it will start outputting deprecation notices for all the places where we are incorrectly passing validator a non-String value.

Validator only validates strings [see here](https://github.com/chriso/validator.js#strings-only).

The warnings output during the tests look like this (there are quite a few):

`validator deprecated you tried to validate a number but this library (validator.js) validates strings only. Please update your code as this will be an error soon. core/server/data/validation/index.js:159:39`

This is an indicator of something we're doing very wrong, the deprecation notices will be very helpful for locating all the places it is happening and getting rid of them. However, I don't think we should upgrade without also fixing the issues, as this will result in lots of extra output in the test/server console that is just confusing if you don't know what it is.

Please note: we use a special convention for the title line in a commit message changing a dependency:

```
deps: validator@x.x.x

closes #6462

- Summary of 
- things that were 
- changed
```

### PR DESCRIPTION
**Pull request by joerx**
Got some code for us? Awesome 🎊!

Please include a description of your change & check your PR against this list, thanks!
- [x] Commit message has a short title & issue references
- [x] Commits are squashed 
- [x] The build will pass (run `npm test`).

More info can be found by clicking the "guidelines for contributing" link above.

closes #6462
- monkey-patch validator.extends() since it was dropped by validator @5.0.0
- coerce input to string prior to validation (custom toString func)
- need to handle boolean validation based on column type not isIn()

### CODE REVIEW

**Comment by joerx** [comment]
validator 5.0.0 breaks validation of non-string input values. this has some interesting
consequences.

* to maintain backwards compat, the safe approach is to manually coerce input using `toString()`
* problematic are validations like 'isIn' with the reference value(s) being non-string - should we consider 0 == '0'? (assuming yes, for compat)
* the schema had a few validations of the form `isIn: [[0, 1, false, true]]` to validate booleans. `toString(true)` would return 'true' which is not in the target collection - I modified validation code to validate boolean cols via `validator.isBoolean()`
* validation for empty values using `validator.isEmpty()` only works when input is converted to string before, since lodash's `_.isEmpty()` behaves unexpected when dealing with booleans and numbers

File: package.json [hunk 1]
```diff
@@ -103,6 +103,16 @@
...
+    "ignore": [
+      "bower",
+      "bluebird",
+      "glob",
+      "lodash",
+      "mysql",
+      "nodemailer",
+      "pg",
+      "showdown-ghost",
+      "validator"
```
Comment (ErisDS) [inline package.json]: This PR can remove this line :smile:  You'll probably need to do a rebase as  I merged a PR where this was split out onto lines earlier today.

**Comment by ErisDS** [comment]
This is looking good thanks :+1:  Having the boolean validations be automatic rather than explicit seems sensible. Did you do much testing to see if there were cases where the assertion/deprecation message gets spat out?

**Comment by joerx** [comment]
Well, it wouldn't create warnings anymore, since `validator` `5.0` it would just throw up and fail miserably. Didn't do much manual testing since I assumed the tests would sort it all out - as long as code is going via `core/server/data/validation` and not using `validator` directly it should be alright. Can do some additional smoke testing tomorrow though, if you want.

**Comment by ErisDS** [comment]
@joerx would be excellent if you could check for any code paths using validator directly (hope not!) and run a couple of tests - perhaps try sending weird values to the API? 

Other than that this looks good to go :)

**Comment by joerx** [comment]
OK, so I did some searching and I actually found quite a few direct uses of 'validate' in the code: 

- api/authentication (email in generateResetToken) -
- config/index (haven't understood config well enough to qualify)
- sitemap/user-generator (imageUrl - can't qualify but looks safe)
- mail/index (from address - loaded from config = safe?)
- models/user (password)
- models/base (sanitize but disabled)
- models/post (via sanitize but disabled)

Manually testing Posts/Settings APIs produced 500s - ironically not related to the validator at all, but caused by `trim()` or `replace()` being called on the `title` and `markdown` properties. I added a few tests and the related fixes - I only tested cases where validator is involved though.

Another thing I noticed: `validator.sanitize()` is not used by models (anymore?) but still required. Maybe some cleanup due?

**Comment by joerx** [comment]
I'm still wondering what's the best option to avoid `'' + foo` - lodash's `toString()` seems to work well, but it's only available in `3.10.1`. :disappointed:

**Comment by ErisDS** [comment]
This looks like it's probably catching and fixing some gnarly edge cases! 

> Another thing I noticed: validator.sanitize() is not used by models (anymore?) but still required. Maybe some cleanup due?

Yes, it's always good to remove dead code, even if it might be brought back later.

> I'm still wondering what's the best option to avoid '' + foo - lodash's toString() seems to work well, but it's only available in 3.10.1. :disappointed:

Reading [this stackoverflow](http://stackoverflow.com/questions/11083254/casting-to-string-in-javascript) on the subject, highlights why `_.toString()` is superior to any of the native versions as it handles nulls and undefined correctly. We're stuck on an older version of lodash atm, but we could include [this module](https://www.npmjs.com/package/lodash.tostring) to get access?

**Comment by joerx** [comment]
Any reason why `lodash` can't be upgraded? Using `lodash.tostring` for the time being. Looks good to merge from my side.

**Comment by ErisDS** [comment]
I left a note on this PR as to why I didn't upgrade it, but I can't say that I remember what the note means :disappointed: :

**Comment by ErisDS** [comment]
Ok, I think this is ready to merge, with one tiny exception - can you please remove the caret from the new dependency version in `package.json`? Thanks!

**Comment by joerx** [comment]
Done

**Comment by ErisDS** [comment]
@joerx This is all looking good, unfortunately all GitHub notifications started going into my spam folder over the weekend and so I didn't see this before it got a merge conflict 😞 

If you have time to rebase that'd be great otherwise I'll look at doing it myself sometime later this week.

**Pull request merged**